### DISCUSSION TITLE
P2P Web integrated with Lokinet

### DISCUSSION THREAD

**Discussion by gennaro-arch** [oxen-io/lokinet#1591] [Ideas]
This is a proposal that I think is very important and it concerns a decentralized World Wide Web. When it comes to Tor darknet sites, almost all of the time they are deanonymized due to server vulnerabilities and not Tor itself, a decentralized and anonymous web would be the solution to increase the degree of reliability of a hidden service by many orders of magnitude.
Being decentralized, all websites can only do client-side scripting and must be programmed using paradigms suitable for decentralization.
Theoretically such a thing already exists and it's called [Zeronet](https://github.com/HelloZeroNet/ZeroNet) and I highly recommend trying it if you don't know it to understand its potential. However, despite its popularity, the project has stalled for years and was developed by a single person. There are problems that can be solved using the blockchain. As in torrents websites are identified with hashes, using LNS we can have a decentralized and comfortable resolution. Then it is not anonymous by default, in fact it leaks a lot of metadata and if you want to be sure you have to use Tor in "always" mode and in any case it is not very efficient. While with Loki's we can use Lokinet by default. Another flaw design from Zeronet is user management. Being decentralized, any data is cryptographically authenticated and in the case of private communication it is encrypted; to do this, each user has a pair of rsa keys and the public key is signed by the authorities. This clearly leads to a non-trustless system, while with Loki we can use the same approach as Session and in addition provide API in javascript to integrate the transfer of funds in one click in the various sites. A customized Firefox setup similar to Tor Browser can provide users with an adequate level of fingerprint protection that is still cost effective compared to the risks of using a centralized service that may have been compromised at any time, not knowing what happens from the server side. Being P2P when a site no longer has peers it disappears forever, maybe we can implement Service Nodes like on Session to manage sites with few peers, but I don't know how sustainable this can be for the network and in any case I don't know if it is necessary.
In any case I think a decentralized web along the lines of Zeronet, but changing the design to be integrated with Lokinet is definitely a breakthrough and deserves to be talked about.

**Comment by majestrate** [comment]
I had a similar idea to this but lack the time to explore it, similar to how the browser is a js/html/http kernel i want to make a lua/html/bittorrent kernel that does p2p offload and p2p message passing using bittorrent v2.

**Reply by gennaro-arch** [reply]
If a browser is used there is already a sandbox done well, but how do you do it with lua?

**Reply by majestrate** [reply]
the idea is to totally rewrite the whole thing from zero for tech debt related reasons.

**Reply by ple1n** [reply]
how about wasm instead lua. 

zeronet is floodsub+sqlite+fixed verification criteria. what should it be for lokinet i wonder ? 

also i'd mention ipfs. it's a pretty good foundation for dweb, but totally no privacy though.

**Comment by KeeJef** [comment]
Arweave for decentralized storage and Lokinet for anonymous access solves most of this IMO, there's already an Arweave gateway on Lokinet, you can find it here http://arweave.f59q8xa7m3bmkfi11oijuh4w77ar49hgx8pdexoqonehcywnifsy.loki/

**Comment by gennaro-arch** [comment]
Using WASM people can compile/implement their favourite DBMS.

Il ven 8 apr 2022, 05:09 planetoryd ***@***.***> ha scritto:

**Reply by ple1n** [reply]
yep, and database could probably be some shared library, and now we need a package manager for wasm

**Comment by mx5kevin** [comment]
Zeronet's system can be well integrated with Lokinet. ZeroNet having huge issues with large files downloading. Downloads are not resumed when the program closed, files are paused, have a preset 10 MB download limit. But can handle up to more than 1500MB large files if this limit are increased. Which most users can't handle. The original developer left the project since at 2019. A functioning download system that torrent clients also use would be needed downloading complete sites with all files, and downloading separated files easy way. Like downloading website with all large files, downloading website with only selected files mods where the file downloads are always completed and without file size limits. Currently, there is no one who would do this. We need a good programmer who is experienced in developing torrent clients. Those who are currently developing the project are not the original developer and just waiting for the money and are not doing anything significant. They are incompetent programmers and professionals. Open source freely editable project, whoever understands it should develop it further. With a full i2p, Lokinet, Tor, VPN support, using 100% anonymous indirect connections easy to implement and integrate. ZeroNet projekt are the best P2P web project from all i know, which is worth dealing with more seriously and protect users anonymity, suitable for decentralized social media platforms, forums and other social media community, and decentralized full feature websites, web hosting.

**Someone who is well versed in the development of torrent clients (download managers), and a good programmer should take over the development of the project. Don't wait for anyone to develop on ZeroNet, the original developer left the project, everyone else who has touched it so far is just a group of botchery who are only waiting for the money.**

**Reply by majestrate** [reply]
zeronet is a locally running web server with no concept of CORS. I don't personally consider it a candidate just because of that and I think the original developer left after he realized how much work would be involved to rectify that.

**Reply by mx5kevin** [reply]
> zeronet is a locally running web server with no concept of CORS. I don't personally consider it a candidate just because of that and I think the original developer left after he realized how much work would be involved to rectify that.

Its developer lives in the same country where I live. He may be afraid that he might have a legal problem’s, some people try to harm him (as Chinese agents). It is also possible that the Hungarian authorities sought him out and scared him. Most of the problems have been solved by the developer. Need a working download system that completes started downloads and is easy to use. If this was done well, the network would already work well. This can be easily fixed by a good developer. For large files, the preset limit of 10 megabytes should be removed, easy way setup the user want the website with all large optional file, or separated want large optional files, allow to +add download links to the client, do not pause the files always download it, run at system startup default way to seed all files. Pause, delete, options in files menu. If these were solved, the system would work perfectly. Most of the problems were solved by the developer. What should be fixed is its download system with a well-functioning download management system that would be worth the work invested in it. If this part works, the other problems related to the network are just small problems and the system would work perfectly. It's not so impossible that it's not worth to fix this a good programmer.

**Comment by mx5kevin** [comment]
I completely agree with the integration of ZeroNet with Lokinet. Among the many networks, the most suitable for the purpose. However, for this its bugs must be corrected.

**Reply by ple1n** [reply]
or https://github.com/freenet/locutus/ instead

**Reply by majestrate** [reply]
freenet made a new implementation? neat.
